src/content/post/20161122_drupal_application_bottlenecks_avoid_with_profiling.md

```diff
@@ -0,0 +1,160 @@
+Obviously, remember to puts on the mysql slow query log.
+
+#### An example:
+
+This is a Blackfire profiling output: [https://blackfire.io/profiles/f911c4ad-792b-4977-af51-d9e3b7649d24/graph](https://blackfire.io/profiles/f911c4ad-792b-4977-af51-d9e3b7649d24/graph)
+We can clearly see a 'red branch' that indicates us where the most of time is spent to generate the page and we can see that the _node_load_ function was called 200 times after the call to _dsc_module_block_view_ , uhm IMHO it's better take a look to this function:
+
+<pre>
+function dsc_module_block_view($delta = '') {
+  $block = array();
+  switch ($delta) {
+    case 'page_block':
+
+    $node_titles = [];
+    for ($i = 1; $i < 200; $i++) {
+      $node = node_load($i);
+      $node_titles[] = $node->title;
+    }
+    $content = '<ul>';
+    foreach ($node_titles as $node_title) {
+      $content .= '<li>;' . $node_title . '</li>';
+    }
+    $content .= '</ul>';
+    $block['subject'] = t('My block');
+    $block['content'] = array(
+      '#type' => 'markup',
+      '#markup' => $content,
+      '#title' => t('Last 100 content'),
+    );
+    break;
+  }
+  return $block;
+}
+</pre>
+
+We have a simple Drupal 7 application that display 200 node's titles, but in the wrong way.
+With a simple refactoring we can improve the block generation:
+
+<pre>
+function dsc_module_block_view($delta = '') {
+  $block = array();
+
+  switch ($delta) {
+    case 'page_block':
+
+      $n_ids = [];
+
+      for ($i = 1; $i > 200; $i++) {
+        $n_ids[] = $i;
+      }
+
+      $results = db_select('node', 'n')
+        ->fields('n', array('title'))
+        ->condition('nid', $n_ids, 'IN')
+        ->execute()
+        ->fetchAll();
+      $content = '<ul>';
+      foreach ($results as $node_title) {
+        $content .= '<li>' . $node_title->title . '</li>';
+      }
+      $content .= '</ul>';
+      $block['subject'] = t('My block');
+      $block['content'] = array(
+        '#type' => 'markup',
+        '#markup' => $content,
+        '#title' => t('Last 100 content'),
+      );
+      break;
+    }
+    return $block;
+  }
+}
+</pre>
+
+And the Blackfire's verdict become [https://blackfire.io/profiles/b7243495-11d8-41fc-a374-e7bfe6e21ab7/graph](https://blackfire.io/profiles/b7243495-11d8-41fc-a374-e7bfe6e21ab7/graph): 1/3 of the time execution saved only with this refactoring.
+
+Obviously this is a very simple example but it's not a remote possibility that these things happens.
+
+### Some pro-tip to write code with an eye on performances
```

Comment:
`### Some pro-tips on how to write code with an eye on performance` mi pare la versione migliore.